test/jdk/java/util/IdentityHashMap/Basic.java

```diff
@@ -38,33 +43,84 @@
  * @run testng Basic
  */
 
-// TODO need this? @modules java.base/java.util:open
-// TODO add tests using null
-// TODO copy constructor instead of k3 and v3 (key and value not in map)?
+// NOTE: avoid using TestNG's assertEquals/assertNotEquals directly on two Map instances,
+// as its logic for testing collections equality is suspect. Use checkContents() to assert
+// that a map's entrySet contains exactly the expected mappings.
+
+// TODO remove(k, v)
+// TODO replace(k, v1, v2)
+// TODO add tests using null keys and values
+// TODO view collections iterators
+// TODO Map.Entry::setValue
 
 public class Basic {
+    /*
+     * Helpers
+     */
+
     record Box(int i) {
         Box(Box other) {
             this(other.i());
         }
     }
 
-    Predicate<Box> isIdenticalBox(Box b1) {
-        return b2 -> b1 == b2;
+    // Checks that a collection contains exactly the given elements and no others, using the
+    // provided predicate for equivalence. Checking is performed both using contains() on the
+    // collection and by simple array searching. The latter is O(N^2) so is suitable only for
+    // small arrays. No two of the given elements can be equivalent according to the predicate.
+
+    @SafeVarargs
+    private <E> void checkContents(Collection<E> c, BiPredicate<E,E> p, E... given) {
+        @SuppressWarnings("unchecked")
+        E[] contents = (E[]) c.toArray();
+
+        assertEquals(c.size(), given.length);
```

Comment:
I believe testng has the expected values in front in the `assertEquals` methods, as embodied in the exception messages, so this should be `assertEquals(given.length, c.size());`. Applies to other places.

Reply:
No, TestNG is `assertEquals(actual, expected)` which is irritatingly the opposite of JUnit.

https://github.com/cbeust/testng/blob/master/testng-asserts/src/main/java/org/testng/asserts/Assertion.java#L151

This will make things quite tedious when/if we convert to JUnit.

Reply:
There’s a reason why I prefer using [AssertJ], where the calls are:
```java
Assertions.assertThat(actual)
	.isEqualTo(expected);
```

[AssertJ]: https://assertj.github.io/doc